### ISSUE TITLE
Data gets corrupted in driver

### ISSUE THREAD

**Issue by king33** [qutheory/mysql#10]
I'm running todobackend tests against my local installation of MySQL.

State of the todos table before tests begin:

`mysql> select * from todos;
Empty set (0.00 sec)
`

First insert into the table looks good, title="walk the dog".  Breaking at getAll and looking at results from a select:

![mysql-first-insert](https://cloud.githubusercontent.com/assets/14813248/17031475/04b5453a-4f3a-11e6-863e-beecc19cdbc5.png)

Then clear is called and `truncate table todos' is run, followed by a second insert of a todo item with the title="blah". 

At another breakpoint in getAll, there's a discrepancy between what is shown in the todos table and what we see in xcode.  Running select from the command line shows that the table is as expected:

<img width="334" alt="correct-table-entry" src="https://cloud.githubusercontent.com/assets/14813248/17031679/f7cc7446-4f3a-11e6-92f9-18db2dc4dc9a.png">

However, results from select in xcode show that the title has been corrupted:

![mysql-second-insert](https://cloud.githubusercontent.com/assets/14813248/17031600/8e65d65a-4f3a-11e6-85bd-168054b56d00.png)

We see this issue when multiple threads are being used.  Our local test suite that runs on a single thread does not expose the problem, so it seems like a concurrency issue.

@rfdickerson

**Comment by tannernelson** [comment]
Thanks for reporting this @king33. Could you paste the code snippets that are leading to this? I'm curious as to whether you are re-using the same connection for each request or spawning new connections. Seeing that code would help a lot.

If at all possible you could fork `qutheory/mysql` and create a failing test case that you submit as a PR, that would help me nip this problem in the bud the quickest.

**Comment by king33** [comment]
Hey, sure, np. I'm creating a new database and connection for every execute and re-using a connection in add to insert and select last_insert_id().

```swift
public func clearAll(oncompletion: (ErrorProtocol?) -> Void) {
        do {
            let query = "TRUNCATE TABLE todos"
            try getDatabase().0?.execute(query)
            oncompletion(nil)
            
        }
        catch {
            Log.error("There was a problem with the MySQL query: \(error)")
            oncompletion(TodoCollectionError.CreationError("There was a problem with the MySQL query: \(error)"))
        }
    }

public func get(withUserID: String?, oncompletion: ([TodoItem]?, ErrorProtocol?) -> Void) {
        let user = withUserID ?? "default"
        
        do {
            let query = "SELECT * FROM todos WHERE owner_id=\"\(user)\""
            let results = try getDatabase().0?.execute(query)
            
            let todos = try parseTodoItemList(results: results!)
            oncompletion(todos, nil)
            
        }
        catch {
            Log.error("There was a problem with the MySQL query: \(error)")
            oncompletion(nil, TodoCollectionError.CreationError("There was a problem with the MySQL query: \(error)"))
        }
    }

public func add(userID: String?, title: String, order: Int, completed: Bool,
                    oncompletion: (TodoItem?, ErrorProtocol?) -> Void ) {
        let user = userID ?? "default"
        
        do {
            let completedValue = completed ? 1 : 0
            
            let query = "INSERT INTO todos (title, owner_id, completed, orderno) VALUES ( \"\(title)\", \"\(user)\", \(completedValue), \(order))"
            //print("\n*** in ADD: query \(query)\n")
            
            let dbConnection = try getDatabase()
            try dbConnection.0?.execute(query, [], dbConnection.1)
            
            let result = try dbConnection.0?.execute("SELECT LAST_INSERT_ID()", [], dbConnection.1)
            
            guard result?.count == 1 else {
                oncompletion(nil, TodoCollectionError.IDNotFound("There was a problem adding a TODO item"))
                return
            }
            
            guard let documentID = (result?[0]["LAST_INSERT_ID()"])?.int
                where documentID > 0 else {
                    oncompletion(nil, TodoCollectionError.IDNotFound("There was a problem adding a TODO item"))
                    return
            }
            let todoItem = TodoItem(documentID: String(documentID), userID: user, order: order, title: title, completed: completed)
            //print("\n*** in ADD: new todo: \(todoItem)\n")
            oncompletion(todoItem, nil)
        }
            
        catch {
            Log.error("There was a problem with the MySQL query: \(error)")
            oncompletion(nil, TodoCollectionError.CreationError("There was a problem with the MySQL query: \(error)"))
        }
    }

    private func parseTodoItemList(results: [[String : MySQL.Value]]) throws -> [TodoItem] {
        var todos = [TodoItem]()
        for entry in results {
            let item: TodoItem = try createTodoItem(entry: entry)
            todos.append(item)
            
        }
        print("\n*** in parse, new todos: \(todos)\n")
        return todos
    }
    
    private func createTodoItem(entry: [String : MySQL.Value]) throws -> TodoItem {
        var tid: Int = 0, user: String = "", title: String = "", orderno: Int = 0, completed: Int = 0
        for(key, value) in entry {
            
            if key == "tid" {
                tid = value.int!
                continue
            }
            if key == "owner_id" {
                user = value.string!
                continue
            }
            if key == "title" {
                title = value.string!
                continue
            }
            if key == "orderno" {
                orderno = value.int!
                continue
            }
            if key == "completed" {
                completed = value.int!
                continue
            }
        }
        let completedValue = completed == 1 ? true : false
        
        let todoItem = TodoItem(documentID: String(tid), userID: user, order: orderno, title: title, completed: completedValue)
        return todoItem
        
    }

private func getDatabase() throws -> (MySQL.Database?, Connection?) {
        
        do{
            let mysql = try MySQL.Database(
                host: self.host,
                user: self.username,
                password: self.password,
                database: self.database
            )
            let connection = try mysql.makeConnection()
            return (mysql, connection)
        } catch {
            Log.error("Failed to create a connection to MySQL database")
        }
        return (nil, nil)
    }
`

**Comment by tannernelson** [comment]
I think this PR should do the trick: https://github.com/qutheory/mysql/pull/11

**Comment by rfdickerson** [comment]
We are getting a crash inside of the defer statement that closes the connection.

<img width="800" alt="screen shot 2016-07-22 at 10 55 00 am" src="https://cloud.githubusercontent.com/assets/1310293/17063094/481c1594-4ffb-11e6-953d-2907cad6162f.png">

This is invoked from our:

<img width="801" alt="screen shot 2016-07-22 at 10 55 12 am" src="https://cloud.githubusercontent.com/assets/1310293/17063104/51ab633a-4ffb-11e6-8e75-f536d81f5ba6.png">

**Comment by tannernelson** [comment]
Are you sure this is related to the fix from #11?

**Comment by rfdickerson** [comment]
It seems to emerge only after we update to the fix. You seeing the same thing, @king33 ?

**Comment by tannernelson** [comment]
The code added in #11 _should_ only affect the parsing of values from `libmysql`. I do not believe it is used at all for the serialization of those values. 

It also looks like the query you pasted is doing manual string interpolation which could be causing that. 

I'm working on an update right now that brings in our `qutheory/core` package. This will make the libmysql <--> swift MySQL conversions more simplified and well tested.

**Comment by king33** [comment]
Hi Tanner, I just wanted to touch base on this. I'm happy to test your update when it's ready. Thank you.

**Comment by tannernelson** [comment]
`0.2.4` should have the update for which this thread was started. Are you seeing a new bug?

**Comment by king33** [comment]
I am seeing the crash that @rfdickerson describes above, it's a separate issue.  We both saw that with 0.2.4.

**Comment by tannernelson** [comment]
Could you create a new issue for it? Then we can see if it is still happening in `0.3.0` which should be released very soon.

**Comment by tannernelson** [comment]
MySQL 0.3 has been released:
https://github.com/qutheory/mysql/releases/tag/0.3.0
